### PR TITLE
[WIP] Resolve cursor timeouts

### PR DESCRIPTION
**Pull request by benjaminpkane**
Work towards resolving #1002 ...

@ehofesmann I'd consider this more of an experiment than anything at this point.

### Background
MongoDB offers a "no timeout" option for typical `find()` cursors, but we use an "aggregation" cursor for sample iteration (and, in general, to support views) which do not have this option.

Our cursors do offer a batch size option. It may be that to many samples are being sent per batch which causes to long of wait between cursor requests.

The details of it are that the cursor may be sending too many samples at a time per cursor request, which leaves the cursor idle for too long.

### TLDR

On this branch, using `dataset_or_view.iter_samples(batch_size=N)` may alleviate the issue in #1002 . Maybe where `N=10` or `N=100`. It's hard to say, but `N=1` might be terribly slow.

### Example

```py
import fiftyone.zoo as foz

for sample in dataset.limit(100).iter_samples(batch_size=10):
    continue
```

### CODE REVIEW

**Comment by benjaminpkane** [comment]
Thinking out loud, we could also just track the current index during iteration and create a new one with a `skip(idx)` stage if the cursor times out...

**Comment by brimoor** [comment]
@benjaminpkane Thanks for taking a stab at some possible fixes for this!

[Looks like](https://docs.mongodb.com/manual/tutorial/iterate-a-cursor/#cursor-batches) MongoDB's default batching strategy is to use a batch size of 101 for the first batch, followed by an undetermined batch subsequently, which is probably as big as possible such that batch is <= 16MB.

Given the 16MB point, don't you think its more likely that the issue is the overall time and not the batching, especially since #1002 reports a failure after ~1 hour?

I think the `skip()` recovery might be a good option here.

If smaller batch sizes were a viable solution, I'd hate to decrease performance in the typical case to avoid problems only in extreme cases (and would rather users not have to manually configure their way out of errors in extreme cases either)

**Comment by benjaminpkane** [comment]
```py
Given the 16MB point, don't you think its more likely that the issue is the overall time and not the batching, especially since #1002 reports a failure after ~1 hour?

I think the skip() recovery might be a good option here.

If smaller batch sizes were a viable solution, I'd hate to decrease performance in the typical case to avoid problems only in extreme cases (and would rather users not have to manually configure their way out of errors in extreme cases either)
```

My thinking was 16MB batches were too big, but I don't really believe that either. Anyways, closing, and will make a PR for `skip` recoveries.
